Initial PR

ofstudio opened the pull request.

ofstudio committed:
add config

ofstudio committed:
add logger

ofstudio committed:
exclude testdata dir

ofstudio committed:
add deps

ofstudio committed:
add luhn checker

ofstudio committed:
add errs

ofstudio committed:
add repo

ofstudio committed:
error handling refactoring

ofstudio committed:
repo error handling refactor

ofstudio committed:
add usecases

ofstudio committed:
add mocks

ofstudio committed:
config refactor

ofstudio committed:
add login and pass validations

ofstudio committed:
remove reqid from app.error

ofstudio committed:
typo

ofstudio committed:
add http handlers

ofstudio committed:
fix error handling

ofstudio committed:
typos

ofstudio committed:
fix varchar limits

ofstudio committed:
fix logging

ofstudio committed:
add description

ofstudio committed:
fix repo namings

ofstudio committed:
update usecases

ofstudio committed:
fix order num http err code

ofstudio committed:
remove DB from repo interface declaration

ofstudio committed:
more tests on usecases

ofstudio committed:
more handlers and tests

ofstudio committed:
typo

ofstudio committed:
add return param to …UpdateFurther

ofstudio committed:
add Integration config and tests

ofstudio committed:
add integrations

ofstudio committed:
add schemas

ofstudio committed:
go mod tidy

ofstudio committed:
fix repo cfg

ofstudio committed:
fix logger

ofstudio committed:
add main

ofstudio committed:
statictest fix

ofstudio committed:
add Authorization response header

ofstudio committed:
fix status code and withdrawal route

ofstudio committed:
HTTP logger tuning

ofstudio committed:
http log refactor

ofstudio committed:
comments

ofstudio committed:
add comments

ofstudio committed:
add cluster compatibility to migrations

ofstudio committed:
add status to withdrawals list

ofstudio committed:
ignore if no update

ofstudio committed:
fix lock further - missed created_at and updated_ad

ofstudio committed:
change env section syntax

ofstudio committed:
add cluster-mode demo

ofstudio committed:
add IntegrationAccrual PollInterval and Timeout to env and cli config

ofstudio committed:
update readme

mr-fard4y commented on cmd/gophermart/main.go:
```diff
@@ -1,3 +1,87 @@
-func main() {}
+import (
+	"context"
+	"github.com/go-chi/chi/v5"
```
отсоортировать импорты и сгруппировать в следующей полседовательности:
- стандартные пакеты
- внешние импорты/пакеты
- внутренние импорты на пакеты проекта
между группой импортов оставлять разделить в виде переноса строки

сортировку произвести во всех файлах проекта

ofstudio replied:
✅ 2ed9df0

mr-fard4y commented on cmd/gophermart/main.go:
```diff
@@ -1,3 +1,87 @@
+func main() {
+	// Контекст для остановки интеграций
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```
в рамках **cmd** оставить лишь импорты пакетов из **internal** и обработку внешних вызовов (прерывания, сигналы системы)

упростить до 
```
srv := server.NewServerByConfig()
if err:= srv.Run(); err != nil {
       log.Fatal().Err(err).Msg("error while starting server")
} 
```

ofstudio replied:
✅  7885fa4

mr-fard4y commented on cmd/gophermart/main.go:
```diff
@@ -1,3 +1,87 @@
+	}
+
+	// Создаём репозиторий, юзкейсы и интеграции
+	repository, err := repo.NewPGXRepo(&cfg.DB, log)
```
логику инициализация сервера перенести в **internal** и инкапсулировать в рамках конструктора
аналогично с описание хэндлеров, middlware и матчем урл-хэндлер

ofstudio replied:
✅  7885fa4

mr-fard4y commented on internal/app/errors.go:
```diff
@@ -0,0 +1,107 @@
+)
+
+// Ошибки пользователя 1100-1199
+var (
```
можно объекдинить в рамках одно **var**
```
var (
     <группа ошибок-1>

     <группа ошибок-2>

     <группа ошибок-3>
)
```

ofstudio replied:
✅  7885fa4

develodisma replied:
✅  876f7a4

ofstudio replied:
✅  876f7a4

mr-fard4y commented on internal/config/default.go:
```diff
@@ -0,0 +1,31 @@
+)
+
+// Default - конфигурационная функция, возвращает конфигурацию по умолчанию.
+func Default(_ *Config) (*Config, error) {
```
конструкторы принято называть New, New[Smth]
название методов - глагольная форма, определяющая описанную внутрилогику, например, handleRoute, doSmth, initDB
существительные используются для определения свойств в структурах или промежуточными переменными

```
func NewDefault() (*Config, error) {
```

ofstudio replied:
✅  7c1ee01

mr-fard4y commented on internal/handlers/auth.go:
```diff
@@ -0,0 +1,107 @@
+//    Authorization: Bearer <JWT token>
+// ...либо
+//    Authorization: <JWT token>
+func (h *Handlers) authMiddleware(next http.Handler) http.Handler {
```
middlware лучше отделить от **handlers** и вынести в рамках отдельного модуля и/или файла

ofstudio replied:
✅  7c1ee01

mr-fard4y commented on internal/handlers/balance.go:
```diff
@@ -0,0 +1,149 @@
+	})
+}
+
+type BalanceHistoryResponse struct {
```
описания схем данных вынести в отдельный модуль/файл и сгруппировать в рамках описания хэндлеров

ofstudio replied:
✅  28d1363

mr-fard4y commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,67 @@
+	}
+}
+
+func (h *Handlers) Routes() chi.Router {
```
изменить название на **InitRoutes()**
методы должны содержать/передавать назначение логики метода в виде глагольной формы
существительные используются для описания параметров структур, промежуточных переменных

ofstudio replied:
✅  8808ee5

mr-fard4y commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,67 @@
+	return r
+}
+
+func decodePlainText(r *http.Request) (string, error) {
```
можно переместить в файл **utils**, т.к. описание логики непосредственно не связано с глобальной логикой сервера

ofstudio replied:
✅  029d441

mr-fard4y commented on internal/integrations/accrual.go:
```diff
@@ -0,0 +1,227 @@
+)
+
+// Accrual - интеграция с системой начисления бонусов
+type Accrual struct {
```
или описание структуры слишком раздуто - UseCases, client
или название структуры не передает его назначение - AccrualClient, AccrualServer

ofstudio replied:
✅  d8a1c7f

mr-fard4y commented on internal/repo/interface.go:
```diff
@@ -0,0 +1,32 @@
+)
+
+// Repo - интерфейс репозитория
+type Repo interface {
```
интерфейс лучше разделить на отдельные интерфейсы по группам действия
итоговое/текущее описание интерфейса **Repository** объединит все остальные - следуем SOLID

ofstudio replied:
✅  da3d3b2

mr-fard4y commented on internal/repo/repo.go:
```diff
@@ -0,0 +1,59 @@
+type PGXRepo struct {
+	db    *sql.DB
+	log   logger.Log
+	stmts []*sql.Stmt
```
в рамках описания структур использовать полный нейминг
сокращения можно использовать в рамках сигнатур методов и/или как описание промежуточных переменных

ofstudio replied:
✅  121967a

mr-fard4y commented on internal/repo/statements.go:
```diff
@@ -0,0 +1,29 @@
+
+// prepareStmts - подготавливает стейтменты к БД.
+// Возвращает список подготовленных стейтментов.
+func prepareStmts(db *sql.DB) ([]*sql.Stmt, error) {
```
использовать полный нейминг

ofstudio replied:
✅  121967a

mr-fard4y commented on internal/repo/statements.go:
```diff
@@ -0,0 +1,29 @@
+
+// registerStmt - добавляет запрос в список запросов к БД.
+// Возвращает индекс запроса в списке.
+func registerStmt(query string) int {
```
использовать полный нейминг

ofstudio replied:
✅  121967a

ofstudio committed:
[review] fix import sort
https://github.com/ofstudio/gophermart-loyalty/pull/1#discussion_r1002526992

ofstudio committed:
refactor app.Error => errs.Error

ofstudio committed:
typo

ofstudio committed:
add comments on repo autostest

ofstudio committed:
[review] simplify main
https://github.com/ofstudio/gophermart-loyalty/pull/1#discussion_r1002532467

ofstudio committed:
[review] group errors declarations
https://github.com/ofstudio/gophermart-loyalty/pull/1#discussion_r1002533281

ofstudio committed:
[review] refactor config constructors name
https://github.com/ofstudio/gophermart-loyalty/pull/1#discussion_r1002534452

ofstudio committed:
move error response render to errs package

ofstudio committed:
[review] extract auth middleware to separate pkg
https://github.com/ofstudio/gophermart-loyalty/pull/1#discussion_r1002603755

ofstudio committed:
[review] move dto's to dto.go

ofstudio committed:
[review] rename Handlers.Routes() to InitRoutes()
https://github.com/ofstudio/gophermart-loyalty/pull/1#discussion_r1002604372

ofstudio committed:
[review] move decodePlainText to helpers.go

ofstudio committed:
[review] integrations naming

ofstudio committed:
[review] repo interface decomposition

ofstudio committed:
[review] statements instead of stmts
https://github.com/ofstudio/gophermart-loyalty/pull/1#discussion_r1002606433
https://github.com/ofstudio/gophermart-loyalty/pull/1#discussion_r1002606446
https://github.com/ofstudio/gophermart-loyalty/pull/1#discussion_r1002606448

ofstudio committed:
update readme

mr-fard4y commented on cmd/gophermart/main.go:
```diff
@@ -1,3 +1,37 @@
+
+	// Запускаем приложение
+	application := app.NewApp(cfg, log)
+	err = application.Start(ctx)
```
можно использовать сокращенную форму
```
if err := app.Start(ctx); err != nil {
}
 ```

ofstudio replied:
[82d6feb](https://github.com/ofstudio/gophermart-loyalty/pull/1/commits/82d6feb5dc1ab9137e83f4b91901c233536237f6)

mr-fard4y commented on internal/app/app.go:
```diff
@@ -0,0 +1,91 @@
+
+// stopServer - остановка HTTP-сервера
+func (a *App) stopServer() {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
числовую переменную вынести в константу/переменную в рамках файла

ofstudio replied:
[82d6feb](https://github.com/ofstudio/gophermart-loyalty/pull/1/commits/82d6feb5dc1ab9137e83f4b91901c233536237f6)

mr-fard4y commented on internal/handlers/balance.go:
```diff
@@ -0,0 +1,117 @@
+		// Если пользователь не найден — возвращаем 500
+		_ = render.Render(w, r, errs.ErrResponseInternal)
+		return
+	} else if err != nil {
```
оставить линейную структуру
```
if errors.Is {
    return
}
if err != nil {
    return
}
```

ofstudio replied:
[82d6feb](https://github.com/ofstudio/gophermart-loyalty/pull/1/commits/82d6feb5dc1ab9137e83f4b91901c233536237f6)

mr-fard4y commented on internal/handlers/operation.go:
```diff
@@ -0,0 +1,267 @@
+		// то возвращаем 200 ОК
+		w.WriteHeader(http.StatusOK)
+		return
+	} else if err != nil {
```
привести к линейной структуре
```
if errors.Is {
    return
}
if err != nil {
    return
}
```

ofstudio replied:
[82d6feb](https://github.com/ofstudio/gophermart-loyalty/pull/1/commits/82d6feb5dc1ab9137e83f4b91901c233536237f6)

mr-fard4y commented on internal/integrations/accrual.go:
```diff
@@ -0,0 +1,229 @@
+	if err == errs.ErrNotFound {
+		a.log.Debug().Msg("accrual operation: nothing to update")
+		return nil
+	} else if err != nil {
```
привести к линейной структуре
```
if errors.Is {
    return
}
if err != nil {
    return
}
```

ofstudio replied:
[82d6feb](https://github.com/ofstudio/gophermart-loyalty/pull/1/commits/82d6feb5dc1ab9137e83f4b91901c233536237f6)

mr-fard4y commented on internal/integrations/accrual.go:
```diff
@@ -0,0 +1,229 @@
+		// Если получили ошибку TooManyRequests, то обновляем тайминги
+		a.adjustPollTiming(err.RetryAfter, err.MaxRPM)
+		return errs.ErrIntegrationTooManyRequests
+	} else if err != nil {
```
привести к линейной структуре
```
if errors.Is {
    return
}
if err != nil {
    return
}
```

ofstudio replied:
[82d6feb](https://github.com/ofstudio/gophermart-loyalty/pull/1/commits/82d6feb5dc1ab9137e83f4b91901c233536237f6)

mr-fard4y commented on internal/integrations/shop-stub.go:
```diff
@@ -0,0 +1,95 @@
+	if err == errs.ErrNotFound {
+		s.log.Debug().Msg("withdrawal operation: nothing to update")
+		return
+	} else if err != nil {
```
привести к линейной структуре
```
if errors.Is {
    return
}
if err != nil {
    return
}
```

ofstudio replied:
[82d6feb](https://github.com/ofstudio/gophermart-loyalty/pull/1/commits/82d6feb5dc1ab9137e83f4b91901c233536237f6)

ofstudio committed:
[review] code style minors
